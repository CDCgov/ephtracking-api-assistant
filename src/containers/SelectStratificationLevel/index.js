import React, { Component } from 'react'
import PropTypes from 'prop-types';
import axios from 'axios';

class SelectStratificationLevel extends Component {
  constructor(props) {
    super(props);
    this.state = {
      options: [],
      value: ''
    };
    this.handleChange = this.handleChange.bind(this);
    this.getOptions = this.getOptions.bind(this);
  }

  componentDidMount() {
    this.getOptions(this.props.contentAreaId, this.props.geographicTypeId);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.measureId !== this.props.measureId || nextProps.geographicTypeId !== this.props.geographicTypeId) {
      this.setState({ value: '' });
      this.getOptions(nextProps.measureId, nextProps.geographicTypeId);
    }
  }

  async getOptions(measureId, geographicTypeId) {
    if (measureId && geographicTypeId) {
      try {
        const response = await axios(`https://ephtracking.cdc.gov/apigateway/api/v1/stratificationlevel/${measureId}/${geographicTypeId}/0`);
        console.log(response);
        this.setState({
          options: response.data
        })
      } catch (error) {
        console.log(error);
      }
    }
  }

  handleChange(event) {
    this.props.handleSelect(event.target.value);
    this.setState({value: event.target.value});
  }

  render() {
    const { value, options } = this.state;
    const disabled = this.props.measureId === null || this.props.geographicTypeId === null;

    const optionsToRender = options.map((item, index) => (
      <option key={index} value={item.id}>{item.name}</option>
    ));
    optionsToRender.unshift([
      <option key="-1" value="" disabled>Select stratification level</option>,
    ]);

    return (
      <div className="field">
        <label className="label">Stratification level</label>
        <div className="control">
          <div className="select">
            <select value={value} onChange={this.handleChange} disabled={disabled}>
              {optionsToRender}
            </select>
          </div>
        </div>
      </div>
    );
  }
}

SelectStratificationLevel.propTypes = {
  handleSelect: PropTypes.func.isRequired,
  measureId: PropTypes.string,
  geographicTypeId: PropTypes.string
};

export default SelectStratificationLevel;
