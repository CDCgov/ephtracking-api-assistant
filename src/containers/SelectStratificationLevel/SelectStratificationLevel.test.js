import React from 'react';
import { shallow } from 'enzyme';
import axios from 'axios';
import MockAdaptor from 'axios-mock-adapter';

import SelectStratificationLevel from './';

describe('SelectIndicator Container', () => {
  const mock = new MockAdaptor(axios);
  const stratificationlevelMock = [
    {
      id: 1,
      name: 'State',
      abbreviation: 'ST',
      geographicTypeId: 1,
      stratificationType: []
    },
    {
      id: 3,
      name: 'State x Age',
      abbreviation: 'ST_AG',
      geographicTypeId: 1,
      stratificationType: [
        {
          id: 3,
          name: 'Age Group',
          abbreviation: 'AG',
          columnName: 'AgeBandId'
        }
      ]
    }
  ];
  const measurestratificationMock = [
    {
      displayName: 'Age Group',
      isDisplayed: true,
      isRequired: false,
      isGrouped: false,
      isStratificationSelectable: false,
      stratificationItem: [
        {
          name: '18 to 24',
          longName: '18 to 24',
          isDefault: false,
          useLongName: false,
          localId: 1
        },
        {
          name: '25 to 34',
          longName: '25 to 34',
          isDefault: false,
          useLongName: false,
          localId: 2
        },
        {
          name: '35 to 44',
          longName: '35 to 44',
          isDefault: false,
          useLongName: false,
          localId: 3
        },
        {
          name: '45 to 54',
          longName: '45 to 54',
          isDefault: false,
          useLongName: false,
          localId: 4
        },
        {
          name: '55 to 64',
          longName: '55 to 64',
          isDefault: false,
          useLongName: false,
          localId: 5
        },
        {
          name: '65 or older',
          longName: '65 or older',
          isDefault: false,
          useLongName: false,
          localId: 6
        }
      ],
      columnName: 'AgeBandId',
      stratificationTypeId: 3
    }
  ];

  it('renders correctly', async () => {
    mock.reset();
    mock
      .onGet(/\/stratificationlevel\//)
      .reply(200, stratificationlevelMock)
      .onGet(/\/measurestratification\//)
      .reply(200, measurestratificationMock);
    const handleSelect = jest.fn();
    // render without contentAreaId
    const wrapper = shallow(
      <SelectStratificationLevel handleSelect={handleSelect} />
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({ measureId: '333', geographicTypeId: '1' });
    await wrapper.instance().getOptions('333', '1');
    expect(wrapper).toMatchSnapshot();
  });

  it('logs an error message on network error', async () => {
    mock.reset();
    mock.onAny().networkError();
    const handleSelect = jest.fn();
    console.error = jest.fn();
    const wrapper = shallow(
      <SelectStratificationLevel handleSelect={handleSelect} />
    );
    await wrapper.instance().getOptions('333', '3');
    expect(console.error).toHaveBeenCalled();
  });

  it('handles stratification level change', () => {
    const handleSelect = jest.fn();
    const wrapper = shallow(
      <SelectStratificationLevel handleSelect={handleSelect} />
    );
    wrapper.instance().setParameterOptions = jest.fn();
    wrapper.setState({ checked: ['AL', 'GA'] });
    wrapper.update();
    wrapper.instance().handleChange({ target: { value: '32' } });
    expect(handleSelect).toHaveBeenCalledWith('32');
    expect(wrapper.state()).toHaveProperty('value', '32');
  });
});
