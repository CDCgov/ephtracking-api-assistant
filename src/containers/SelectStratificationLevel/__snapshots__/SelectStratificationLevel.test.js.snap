// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectIndicator Container renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectStratificationLevel
    handleSelect={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          className="label"
>
          Stratification level
</label>,
        <div
          className="control"
>
          <div
                    className="select"
          >
                    <select
                              disabled={false}
                              onChange={[Function]}
                              value=""
                    >
                              <option
                                        disabled={true}
                                        value=""
                              >
                                        Select stratification level
                              </option>
                    </select>
          </div>
</div>,
        false,
      ],
      "className": "field",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Stratification level",
          "className": "label",
        },
        "ref": null,
        "rendered": "Stratification level",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="select"
>
            <select
                        disabled={false}
                        onChange={[Function]}
                        value=""
            >
                        <option
                                    disabled={true}
                                    value=""
                        >
                                    Select stratification level
                        </option>
            </select>
</div>,
          "className": "control",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <select
              disabled={false}
              onChange={[Function]}
              value=""
>
              <option
                            disabled={true}
                            value=""
              >
                            Select stratification level
              </option>
</select>,
            "className": "select",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                Array [
                  <option
                    disabled={true}
                    value=""
>
                    Select stratification level
</option>,
                ],
              ],
              "disabled": false,
              "onChange": [Function],
              "value": "",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "-1",
                "nodeType": "host",
                "props": Object {
                  "children": "Select stratification level",
                  "disabled": true,
                  "value": "",
                },
                "ref": null,
                "rendered": "Select stratification level",
                "type": "option",
              },
            ],
            "type": "select",
          },
          "type": "div",
        },
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            className="label"
>
            Stratification level
</label>,
          <div
            className="control"
>
            <div
                        className="select"
            >
                        <select
                                    disabled={false}
                                    onChange={[Function]}
                                    value=""
                        >
                                    <option
                                                disabled={true}
                                                value=""
                                    >
                                                Select stratification level
                                    </option>
                        </select>
            </div>
</div>,
          false,
        ],
        "className": "field",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Stratification level",
            "className": "label",
          },
          "ref": null,
          "rendered": "Stratification level",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="select"
>
              <select
                            disabled={false}
                            onChange={[Function]}
                            value=""
              >
                            <option
                                          disabled={true}
                                          value=""
                            >
                                          Select stratification level
                            </option>
              </select>
</div>,
            "className": "control",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <select
                disabled={false}
                onChange={[Function]}
                value=""
>
                <option
                                disabled={true}
                                value=""
                >
                                Select stratification level
                </option>
</select>,
              "className": "select",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  Array [
                    <option
                      disabled={true}
                      value=""
>
                      Select stratification level
</option>,
                  ],
                ],
                "disabled": false,
                "onChange": [Function],
                "value": "",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "-1",
                  "nodeType": "host",
                  "props": Object {
                    "children": "Select stratification level",
                    "disabled": true,
                    "value": "",
                  },
                  "ref": null,
                  "rendered": "Select stratification level",
                  "type": "option",
                },
              ],
              "type": "select",
            },
            "type": "div",
          },
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`SelectIndicator Container renders correctly 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectStratificationLevel
    geographicTypeId="1"
    handleSelect={[Function]}
    measureId="333"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          className="label"
>
          Stratification level
</label>,
        <div
          className="control"
>
          <div
                    className="select"
          >
                    <select
                              disabled={false}
                              onChange={[Function]}
                              value=""
                    >
                              <option
                                        disabled={true}
                                        value=""
                              >
                                        Select stratification level
                              </option>
                    </select>
          </div>
</div>,
        false,
      ],
      "className": "field",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Stratification level",
          "className": "label",
        },
        "ref": null,
        "rendered": "Stratification level",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="select"
>
            <select
                        disabled={false}
                        onChange={[Function]}
                        value=""
            >
                        <option
                                    disabled={true}
                                    value=""
                        >
                                    Select stratification level
                        </option>
            </select>
</div>,
          "className": "control",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <select
              disabled={false}
              onChange={[Function]}
              value=""
>
              <option
                            disabled={true}
                            value=""
              >
                            Select stratification level
              </option>
</select>,
            "className": "select",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                Array [
                  <option
                    disabled={true}
                    value=""
>
                    Select stratification level
</option>,
                ],
              ],
              "disabled": false,
              "onChange": [Function],
              "value": "",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "-1",
                "nodeType": "host",
                "props": Object {
                  "children": "Select stratification level",
                  "disabled": true,
                  "value": "",
                },
                "ref": null,
                "rendered": "Select stratification level",
                "type": "option",
              },
            ],
            "type": "select",
          },
          "type": "div",
        },
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            className="label"
>
            Stratification level
</label>,
          <div
            className="control"
>
            <div
                        className="select"
            >
                        <select
                                    disabled={false}
                                    onChange={[Function]}
                                    value=""
                        >
                                    <option
                                                disabled={true}
                                                value=""
                                    >
                                                Select stratification level
                                    </option>
                        </select>
            </div>
</div>,
          false,
        ],
        "className": "field",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Stratification level",
            "className": "label",
          },
          "ref": null,
          "rendered": "Stratification level",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="select"
>
              <select
                            disabled={false}
                            onChange={[Function]}
                            value=""
              >
                            <option
                                          disabled={true}
                                          value=""
                            >
                                          Select stratification level
                            </option>
              </select>
</div>,
            "className": "control",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <select
                disabled={false}
                onChange={[Function]}
                value=""
>
                <option
                                disabled={true}
                                value=""
                >
                                Select stratification level
                </option>
</select>,
              "className": "select",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  Array [
                    <option
                      disabled={true}
                      value=""
>
                      Select stratification level
</option>,
                  ],
                ],
                "disabled": false,
                "onChange": [Function],
                "value": "",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "-1",
                  "nodeType": "host",
                  "props": Object {
                    "children": "Select stratification level",
                    "disabled": true,
                    "value": "",
                  },
                  "ref": null,
                  "rendered": "Select stratification level",
                  "type": "option",
                },
              ],
              "type": "select",
            },
            "type": "div",
          },
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
